pkg/apis/pipeline/v1beta1/param_types.go

```diff
@@ -141,17 +141,40 @@ type ArrayOrString struct {
 
 // UnmarshalJSON implements the json.Unmarshaller interface.
 func (arrayOrString *ArrayOrString) UnmarshalJSON(value []byte) error {
+	// ArrayOrString is used for Results Value as well, the results can be any kind of
+	// data so we need to check if it is empty.
+	if len(value) == 0 {
+		arrayOrString.Type = ParamTypeString
+		return nil
+	}
 	switch value[0] {
 	case '[':
+		// We're trying to Unmarshal to []string, but for cases like []int or other types
+		// of nested array which we don't support yet, we should continue and Unmarshal
+		// it to String. If the Type being set doesn't match what it actually should be,
+		// it will be captured by validation in reconciler's validation.
 		arrayOrString.Type = ParamTypeArray
-		return json.Unmarshal(value, &arrayOrString.ArrayVal)
+		if err := json.Unmarshal(value, &arrayOrString.ArrayVal); err == nil {
+			return nil
+		}
 	case '{':
 		arrayOrString.Type = ParamTypeObject
-		return json.Unmarshal(value, &arrayOrString.ObjectVal)
+		if err := json.Unmarshal(value, &arrayOrString.ObjectVal); err == nil {
+			return nil
+		}
 	default:
 		arrayOrString.Type = ParamTypeString
-		return json.Unmarshal(value, &arrayOrString.StringVal)
+		if err := json.Unmarshal(value, &arrayOrString.StringVal); err == nil {
+			return nil
+		}
 	}
+	// if failed to unmarshal to above types we will convert the value to string and
+	// marshal it to string
+	arrayOrString.Type = ParamTypeString
+	arrayOrString.StringVal = string(value)
+	arrayOrString.ArrayVal = nil
+	arrayOrString.ObjectVal = nil
+	return nil
```

Comment:
This function does look weird. Can we avoid using `switch`:

```
	if value[0] == '[' {
		// We're trying to Unmarshal to []string, but for cases like []int or other types
		// of nested array which we don't support yet, we should continue and Unmarshal
		// it to String. If the Type being set doesn't match what it actually should be,
		// it will be captured somewhere else.
		// if failed to unmarshal to array, we will convert the value to string and marshal it to string
		if err := json.Unmarshal(value, &arrayOrString.ArrayVal); err == nil {
			arrayOrString.Type = ParamTypeArray
			return nil
		}
	}
	if value[0] == '{' {
		// if failed to unmarshal to map, we will convert the value to string and marshal it to string
		if err := json.Unmarshal(value, &arrayOrString.ObjectVal); err == nil {
			arrayOrString.Type = ParamTypeObject
			return nil
		}
	}
	arrayOrString.ArrayVal = nil
	arrayOrString.ObjectVal = nil
	arrayOrString.Type = ParamTypeString
	arrayOrString.StringVal = string(value)
```

Also, the unmarshalling does set the `arrayOrString.ArrayVal` to an array of equal length of the value in case of failure. And that is the reason we need to set it to nil explicitly. Will be great if we find a way unmarshal does not change the `ArrayVal` but not a blocker for this PR.

Reply:
I changed it to if, and the json.Unmarshal(value, &arrayOrString.StringVal) is needed, we cannot directly call arrayOrString.StringVal = string(value).

Yes, it is a bit weird that we need to set it to nil since we tried to unmarshal to array and it will create an empty array. It will be worth figuring out a better idea of this as you suggested